src/Microsoft.Azure.SignalR.Common/ServiceConnections/ClientConnectionScopeInternal.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.SignalR.Common.Utilities;
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.Azure.SignalR
+{
+    /// <summary>
+    /// Represents a disposable scope able to carry connection properties along with the execution context flow
+    /// </summary>
+    /// Only allows to carry one copy of connection properties regardless of how many nested scopes are created
+    internal class ClientConnectionScopeInternal : IDisposable
+    {
+        private bool _needCleanup;
+
+        internal ClientConnectionScopeInternal() : this(default)
+        {
+        }
+
+        protected internal ClientConnectionScopeInternal(ClientConnectionScopeProperties properties)
+        {
+            if (ScopePropertiesAccessor<ClientConnectionScopeProperties>.Current == null)
```

Comment:
One question, one can override the behavior by setting `ScopePropertiesAccessor<ClientConnectionScopeProperties>.Current` from outside this scope? Is this the concern leading to a nested ClientConnectionScopeProperties class? Is making `ClientConnectionScopeProperties` private and let ctor construct the properties better?

Reply:
Yes, setting `ScopePropertiesAccessor<ClientConnectionScopeProperties>.Current` would allow nested contexts :( (although this is arguably abuse)  Making `ClientConnectionScopeProperties` private will enforce the non-nestable nature of this class. However it would also force us to add explicit API to change the current values carried by the only copy of `ClientConnectionScopeProperties`. Perhaps this is a better way. Will update the PR shortly!